x/ibc/04-channel/keeper/handshake.go

```diff
@@ -0,0 +1,368 @@
+	if connection.State == ics03types.NONE {
+		return "", errors.New("connection is closed")
+	}
+
+	// TODO: Blocked - ICS05 Not implemented yet
+	// port, found := k.portKeeper.GetPort(ctx, portID)
+	// if !found {
+	// 	return errors.New("port not found") // TODO: ics05 sdk.Error
+	// }
+
+	// if !k.portKeeper.AuthenticatePort(port.ID()) {
+	// 	return errors.New("port is not valid") // TODO: ics05 sdk.Error
+	// }
+
+	channel := types.NewChannel(types.INIT, order, counterparty, connectionHops, version)
+	k.SetChannel(ctx, portID, channelID, channel)
+
+	key := "" // TODO: generate key
+	k.SetChannelCapability(ctx, portID, channelID, key)
+	k.SetNextSequenceSend(ctx, portID, channelID, 1)
+	k.SetNextSequenceRecv(ctx, portID, channelID, 1)
+
+	return key, nil
+}
+
+// ChanOpenTry is called by a module to accept the first step of a channel opening
+// handshake initiated by a module on another chain.
+func (k Keeper) ChanOpenTry(
+	ctx sdk.Context,
+	order types.ChannelOrder,
+	connectionHops []string,
+	portID,
+	channelID string,
+	counterparty types.Counterparty,
+	version,
+	counterpartyVersion string,
+	proofInit ics23.Proof,
+	proofHeight uint64,
+) (string, error) {
+
+	if len(connectionHops) != 1 {
+		return "", types.ErrInvalidConnectionHops(k.codespace)
+	}
+
+	_, found := k.GetChannel(ctx, portID, channelID)
+	if found {
+		return "", types.ErrChannelExists(k.codespace)
+	}
+
+	// TODO: Blocked - ICS05 Not implemented yet
+	// port, found := k.portKeeper.GetPort(ctx, portID)
+	// if !found {
+	// 	return errors.New("port not found") // TODO: ics05 sdk.Error
+	// }
+
+	// if !k.portKeeper.AuthenticatePort(port.ID()) {
+	// 	return errors.New("port is not valid") // TODO: ics05 sdk.Error
+	// }
+
+	connection, found := k.connectionKeeper.GetConnection(ctx, connectionHops[0])
+	if !found {
+		return "", ics03types.ErrConnectionNotFound(k.codespace)
+	}
+
+	if connection.State != ics03types.OPEN {
+		return "", errors.New("connection is not open")
+	}
+
+	// NOTE: this step has been switched with the one below to reverse the connection
+	// hops
+	channel := types.NewChannel(types.OPENTRY, order, counterparty, connectionHops, version)
+
+	// expectedCounterpaty is the counterparty of the counterparty's channel end
+	// (i.e self)
+	expectedCounterparty := types.NewCounterparty(portID, channelID)
+	expectedChannel := types.NewChannel(
+		types.INIT, channel.Ordering, expectedCounterparty,
+		channel.CounterpartyHops(), channel.Version,
+	)
+
+	bz, err := k.cdc.MarshalBinaryLengthPrefixed(expectedChannel)
+	if err != nil {
+		return "", errors.New("failed to marshal expected channel")
+	}
+
+	if !k.connectionKeeper.VerifyMembership(
+		ctx, connection, proofHeight, proofInit,
+		types.ChannelPath(counterparty.PortID, counterparty.ChannelID),
+		bz,
+	) {
+		return "", types.ErrInvalidCounterpartyChannel(k.codespace)
+	}
+
+	k.SetChannel(ctx, portID, channelID, channel)
+
+	key := "" // TODO: generate key
```

Comment:
What is required for key generation here?

Reply:
See response above (should be identical to store key mechanism)